### ISSUE TITLE
valid-jsdoc + preferType warns about "Array" type for "Type[]" syntax

### ISSUE THREAD

**Issue by bwrrp** [eslint/eslint#7533]
**Tell us about your environment**

* **ESLint Version:**
3.9.1
* **Node Version:**
6.8.0
* **npm Version:**
3.10.8

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**

```
{
	"parserOptions": {
		"ecmaVersion": 6,
		"sourceType": "module",
		"ecmaFeatures": {
			"experimentalObjectRestSpread": true,
			"jsx": true
		}
	},

	"env": {
		"amd": true,
		"browser": true,
		"es6": true,
		"mocha": true,
		"node": true
	},

	"globals": {
		"chai": true,
		"sinon": true
	},

	"plugins": [ "react" ],

	"extends": [ "eslint:recommended", "plugin:react/recommended" ],

	"rules": {
		// rules other than valid-jsdoc omitted for clarity

		// Check JSDoc comments
		"valid-jsdoc": ["warn", {
			"prefer": {
				"arg": "param",
				"argument": "param",
				"returns": "return",
				"virtual": "abstract"
			},
			"preferType": {
				"*": "any",
				"Boolean": "boolean",
				"Number": "number",
				"String": "string",
				"Array": "Type[]",
				"array": "Type[]",
				"Object": "{key: Type}",
				"object": "{key: Type}",
				"Function": "function(Type, Type): Type",
				"function": "function(Type, Type): Type"
			},
			"requireReturn": false,
			"requireParamDescription": false,
			"requireReturnDescription": false
		}]
	}
}
```

**What did you do? Please include the actual source code causing the issue.**

On a large codebase, we've been using the Type[] convention in jsdoc-like comments to indicate an array of Type. A common mistake is for developers to use "Array" instead, forgetting to document type type of item in the array. I've been trying to use the valid-jsdoc rule and its preferType option to detect this behavior and suggest a change. A similar approach works fine for objects (use "{key: Type}" instead of "Object"), but I'm seeing some weird behavior when trying to do the same for arrays:

```js
function View () {
    // ...
}

/**
 * @param  {Record[]}  records
 */
View.prototype.update = function (records) {
    this._records = records;
    // ...
};
```

**What did you expect to happen?**
I expected no linter warning to be generated for the code sample above.

**What actually happened? Please include the actual, raw output from ESLint.**

eslint acts as if I typed `@param  {Array}  records`:

```
  5:1  warning  Use 'Type[]' instead of 'Array'  valid-jsdoc

✖ 1 problem (0 errors, 1 warning)
```

While the Type[] syntax may not be the preferred way to indicate array types in jsdoc comments, I did not expect the detected type to change this way.

**Comment by bwrrp** [comment]
Digging through the code, it appears that the code causing this seems to have been introduced to handle the `Array.<Type>` syntax, which does at least include "Array". Unfortunately, doctrine parses both `Type[]` and `Array.<Type>` to the same AST:

```js
JSON.stringify(require("doctrine").parse(`/**
 * @param  {Type1[]}  types1
 * @param  {Array.<Type2>}  types2
 */`, {unwrap: true}), null, '\t')
```

```json
{
	"description": "",
	"tags": [
		{
			"title": "param",
			"description": null,
			"type": {
				"type": "TypeApplication",
				"expression": {
					"type": "NameExpression",
					"name": "Array"
				},
				"applications": [
					{
						"type": "NameExpression",
						"name": "Type1"
					}
				]
			},
			"name": "types1"
		},
		{
			"title": "param",
			"description": null,
			"type": {
				"type": "TypeApplication",
				"expression": {
					"type": "NameExpression",
					"name": "Array"
				},
				"applications": [
					{
						"type": "NameExpression",
						"name": "Type2"
					}
				]
			},
			"name": "types2"
		}
	]
}
```

**Comment by nzakas** [comment]
I'd suggest creating a custom rule to do what you need. 
Unfortunately, it looks like consensus couldn't be reached on this issue and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after 21 days tend never to reach consensus, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.

I'd suggest creating a custom rule to do what you need.
